### DISCUSSION TITLE
how do you use table.unpack?

### DISCUSSION THREAD

**Discussion by subv0x** [RomanKubiak/ctrlr#559] [Q&A] [answered]
how do you use unpack or table.unpack ? (does it work in Ctrlr?)

**Comment by dnaldoog** [comment, marked as answer]
> how do you use unpack or table.unpack ? (does it work in Ctrlr?)

I think `table.unpack()` is only in lua 5.2 so you need to use `unpack()` because Ctrlr uses lua 5.1 ;)

**Comment by subv0x** [comment]
i found this:

http://lua-users.org/lists/lua-l/2013-10/msg00534.html

> The Lua website says that "Function unpack was moved into the table library and therefore must be called as table.unpack." I have Lua 5.2.2 and 'unpack' still works for me. Did they keep it around for back compatibility? Which one should I use?

`unpack` is only present in 5.2 of it's compiled with 5.1 compatibility enabled (i.e. LUA_COMPAT_UNPACK or LUA_COMPAT_ALL were defined).

If you're writing solely for 5.2, use `table.unpack`. If you're writing for 5.1 and 5.2, you can use `local unpack = unpack or table.unpack` at the top of the file and then call it as `unpack`.

so '''unpack()''' how do we use that?

**Reply by dnaldoog** [reply]
Your compiled version of Ctrlr uses lua 5.2???

**Reply by subv0x** [reply]
ok i got it working :-) wasn't doing what i expected yesterday.

useful for quick assign of variables to a table:
```
local unpack = unpack or table.unpack
	local p1,p2,p3,p4 = unpack(p)
	local l1,l2,l3,l4 = unpack(levels)
```

**Reply by subv0x** [reply]
now... is there a `pack(table, item?)` eg: to correspond to `table.pack( table, item?)` - been looking at various` table. etc.` functions.

**Reply by dnaldoog** [reply]
What do you mean by table.pack?

Do you mean add variables or values into a table? 

You would need table.insert() to do that:

I’m not aware of a function like:

t={}
a,b,c=table.pack(t)

**Reply by dnaldoog** [reply]
You could write your own. Something like:

<pre>
local t={2,3,4}
local a,b,c=unpack(t)

function pack(...)
    local t = {}
    for i, v in ipairs(arg) do
        t[i] = v
    end
    return t
end

local t1=pack(a,b,c)
</pre>


[table pack.zip](https://github.com/RomanKubiak/ctrlr/files/11865187/table.pack.zip)

**Reply by subv0x** [reply]
i saw table.pack when i was looking at `unpack( ...)` - which was something chatGPt suggested :D - 
so i did a quick check of references online. obviously `table.insert()` but can you do a straight array-insert with something like 
(a,b,c,d) table.insert (table) ? or you have to iterate? 

i like `unpack(table)` anyway.

asked chatGPT for a list of table. functions and got:

    table.insert(table, pos, value): Inserts value at position pos in table, shifting any elements after pos to higher indices.
    table.remove(table, pos): Removes and returns the element at position pos from table, shifting any elements after pos to lower indices.
    table.concat(table, sep): Concatenates all elements of table into a string, with an optional separator sep between each element.
    table.sort(table, comp): Sorts the elements of table in ascending order, using an optional comparison function comp.
    table.unpack(table): Returns all elements of table as individual values. (Lua 5.2 and later, use unpack() for Lua 5.1)
    table.copy(table): Creates a shallow copy of table.
    table.deepcopy(table): Creates a deep copy of table, including nested tables.
    table.getn(table): Returns the length of table. (Deprecated in Lua 5.2; use #table instead)
    table.foreach(table, func): Iterates over key-value pairs in table, calling func(key, value) for each pair.

any more?

**Reply by dnaldoog** [reply]
You have to iterate or use the force!

I'll bet that function I wrote is super fast anyway!

🤣
